Limit the time passed in transport check loop

hardening opened the pull request:
This patch make it possible to limit the time that is passed when we call
XXX_check_fds functions. This should smooth the treatment between handling inputs
and handling incoming bitmap updates.
The default maximum time is set to 100 ms.

HenryJacques commented:
@hardening would it make sense to be able to put the max time value in the command line?

hardening commented:
@HenryJacques yes we could add something like this and probably remove the async input...

realjiangms commented:
This fixes the issue we discussed in #2338. I found if we use a thin client which has poor CPU and use a complex codec, e.g. rfx, the client will be put in a 'stuck' status where the keyboard and mouse input never get a chance to be handled. Yes, in worst case, it never get handled because each time it consumes a packet and finishes rendering, a new packet arrives. So transport_check_fds() never returns.

I tried to use async-input, it made input event handled but also cause X11 deadlock. Then I don't get further more solution.

The reason is that we have 'for (;;)'  {now it is while(!freerdp_shall_disconnect(transport->context->instance)) } in libfreerdp/core/transport.c.
I checked history and found it was added at 2014-03-10, but I basically can't figure out why. Following is the reason in the comment (used to be):

```
/**
         * Loop through and read all available PDUs.  Since multiple
         * PDUs can exist, it's important to deliver them all before
         * returning.  Otherwise we run the risk of having a thread
         * wait for a socket to get signaled that data is available
         * (which may never happen).
         */
        for (;;)
       {
                /**
                 * Note: transport_read_pdu tries to read one PDU from

```

Interesting the comment is removed. But it doesn't mean the concern disappears.

@hardening do you know why we introduce infinite loop in transport_check_fds() before? I think we might need to be careful and MAKE SURE that the old concern in the old removed comment is no longer available. Or we might have regression, although I don't understand how it run into the risk described in the above comment

This fix itself looks good. I hope we can resolve it this time. This issue has been bothering me for a year

hardening commented:
@realjiangms well you're definitely right and something is missing in my PR.

long story: in the client we WaitForMultipleObject and then we call the check_fds function. In the case of a TLS security connection, we(I) have activated the readahead setting of openSSL that makes a SSL_read call read 64k bytes (IIRC) instead of reading 4 bytes (the size of the record) and then the record itself. The consequence is that we get notified of available bytes on the socket once, a 64k block is read, but if we don't read all the PDUs, we won't get notified again that some bytes are buffered for the socket until some activity happens on the socket. So my patch is missing an event that should act as a "read again in the socket" signal.

Looking at this

realjiangms commented:
@hardening. Thanks for the explanation. I now understand why we need infinite loop when reading.
Looking forward your next update :+1:

bmiklautz commented:
@hardening did I understand correctly that you'll update this PR?

hardening commented:
@bmiklautz yes I will, just a matter of hours.

BUGgs commented:
Hi, any updates on this pull ?

hardening commented:
@BUGgs yeah I have this lying in a local branch, I still need to do some final tests.

hardening commented:
About the test, I feel like the TestConnect will fail if 2 of them run in parallel on the same host (as they try to bind 0.0.0.0:3389). @freerdp-bot can you test again to see if I'm right ?

hardening commented:
@freerdp-bot don't you wanna test again ?

akallabeth commented:
@freerdp-bot test

hardening commented:
I have added some changes so that TestConnect has less chance to fail.

akallabeth commented on libfreerdp/core/test/TestConnect.c:
```diff
@@ -241,19 +276,23 @@ static int testSuccess(void)
 {
+	int randomPort;
+
+	randomPort = 3389 + (random() % 20);
```
@hardening Good idea to use a random port, but maybe it would be even better if we spread the port range a little more (and maybe check availability first to avoid collisions with running services)

ropetry commented:
Great. i will try to test as soon as you give a go 8-)..

realjiangms commented on libfreerdp/core/transport.c:
```diff
@@ -864,8 +895,15 @@ int transport_check_fds(rdpTransport* transport)
+		now = GetTickCount();
 	}
 
+	if (now > dueDate)
```
Do we miss the event if the loop exits when "now == dueDate"?

realjiangms commented:
The fix itself looks good to me. The new introduced reread event resolves the issue we discussed before and we no longer need to 'read all PDU' when we receive socket event.

But some code which depends on GetEventHandle seems broken. They don't get reread event.
Maybe we should discard the GetEventHandle API to avoid wrong usage. Also all code with BIO_GET_EVENT need to be reviewed.

Above concern could be handled in another separate PR.

Actually our real problem is that the frontBio event doesn't match our requirement. A prefect event we need is one which is always 'set' when there's still something to be read and always 'reset' when there's nothing to be read. However we use event from the most bottom layer in our tcp/tls stack, so it can't reflect upper layer status correctly. And unfortunately the linux version of event is built directly on socket fd and we basically can't 'set' it manually even if we know 'there is still something to be read'

ropetry commented:
Hi, is there a way to download a "patched" source, oder just to download this pull request as a diff so that i can apply it by myself? i can´t find a "download" option for the patches or source. perhaps i missed it in the github gui 8-(..
o.k found it.. just apply a .diff to the pullreques id.. i keep this here for others..
https://github.com/FreeRDP/FreeRDP/pull/3364.diff or .patch

thanks
roman

ropetry commented:
Hello, anyone around with some news about the patch ?
@akallabeth @hardening

hardening commented:
Ok I have updated the patch according to the comments and rebased it too, so review is welcome.

@realjiangms I agree that GetEventHandle should be removed from the public API

akallabeth commented on libfreerdp/core/transport.c:
```diff
@@ -1075,19 +1112,26 @@ rdpTransport* transport_new(rdpContext* context)
 	    || transport->connectedEvent == INVALID_HANDLE_VALUE)
 		goto out_free_receivebuffer;
 
+	transport->rereadEvent = CreateEvent(NULL, TRUE, FALSE, NULL);
```
@hardening Missing event close in transport_free

hardening replied:
Yep I've seen that too yesterday, patch coming

akallabeth commented:
@hardening Could you rebase this pull and maybe run ```./scripts/format_code.sh``` on every changed file in this pull to get a consistent style on master?

hardening commented:
I did some changes in the astyle parameters, the most controversial change must be to set the line width to 100: who still use a 80 column terminal to code these days ? (and BTW when formatting is done by humans that's the kind of limit we have in the code)
I have applied that on my modified files. If it's too intrusive, I will revert and apply the original style.

akallabeth commented:
@hardening Fine with that, but we should not do that too often ;)

akallabeth commented:
@freerdp-bot test please

bmiklautz commented:
@hardening may I suggest we also add a command line parameter that this can be configured? With 0 to disable it completely?

ropetry commented:
@hardening @bmiklautz 
+1 for the command line parameter ...

hardening commented:
@ropetry @bmiklautz you mean that 0 would restore the old behaviour, where we treat all the packet we can grab, with the potential bug that will never exit that loop ?

bmiklautz commented:
@hardening I'd think so. If you don't agree 0 needs to be handles properly at least ;).

realjiangms commented:
+1 for command line parameter and 0 for old behavior. Actually I think it should be default behavior to be backward compatible. By the way, why do you define 100ms as default value? Is it practically best as default?

hardening commented:
@bmiklautz I don't really see why we would keep the old behaviour. "Being compatible" doesn't really make sense. You can mimick the old behaviour with setting a big value, but I really don't see why somebody would choose to do so and reduce the responsiveness...

@realjiangms  the 100 ms was choosen by the intuition that it was the maximum acceptable value to handle user inputs. Anyway if someone has a better number I'm ok with it.

bmiklautz commented:
@hardening you are right. A large value can still "restore" the old behavior but what happens if 0 is passed via the command line parameter?

hardening commented:
@bmiklautz well for now no command line parameter is available for that, so...

hardening commented:
Added a command line parameter.

ropetry commented:
@hardening @bmiklautz i think this patch has introduced a new problem. please see issue 3628
if i set the max-loop-time to 1000 instead of the default 100 the session ( with powerpoint an dother applications) run smoother and not has minimal lags.

realjiangms commented:
@hardening The problem from @ropetry makes it really interesting :)
 "Being compatible" does really make sense. He is somebody would choose to do so and reduce the responsiveness, but makes server generates less drawing command and works better :)
